### ISSUE TITLE
In softmax layer of word2vec, do we use cosine similarity or dot product?

### ISSUE THREAD

**Issue by tranquil-coder** [RaRe-Technologies/gensim#3008]
#### Problem description

I have read the paper "Efficient Estimation of Word Representation in Vector Space".  This article says that, we use **cosine similarity**  in softmax layer of word2vec. But someone says that gensim uses **dot product** in softmax layer of word2vec while uses **cosine similarity** between word vectors which have been trained. I have not read the source code, and I wanted to confirm  whether use  dot product in softmax layer and use cosine similarity after trained.

**Comment by gojomo** [comment]
Word2vec implementations tend not to use plain softmax. Going back to the `word2vec.c` code released by Google researchers, they use either 'hierarchical softmax' or 'negative sampling' – each of which require far less calculation that true softmax. 

I can't find any reference to using 'cosine similarity' in the "Efficient Estimation of Word Representation in Vector Space" paper. Searching for the word 'cosine' only reveals one usage, paired as 'cosine distance', and in the context of using a model's final word-vectors, not in the context of training. So I'm not sure what you're referring to. 

The Gensim `Word2Vec` class is a faithful re-implementation of the algorithm as demonstrated in Google's released `word2vec.c` code. You can review the source code if you need deep details, but there is no discrepancy with `word2vec.c` or other prominent implementations. 

Questions are best asked on the project discussion list, at <https://groups.google.com/u/1/g/gensim>, rather than in this bug/feature-request tracker.
